Automattic/woocommerce-payments
# Cloak client secret

Fixes #4774

#### Changes proposed in this Pull Request

This PR includes a PoC about hiding the exposed `client_secret` variable on the  `create_payment_intent` and `create_setup_intent` responses from automated scripting for card testing attacks on the client. 

Basically, it encrypts the `client_secret` variable with AES-128-CBC encryption just before pouring it to the browser, and decrypts it just before the calls to the Stripe API via the Stripe JS library. 

This is a measure to prevent the attacker reaching to the `client_secret` variable, which lets them call Stripe REST API directly to confirm intents, without using our server endpoints.

Also it can be enabled/disabled by clicking the "**Enable public key encryption**" checkbox in **Payments > Settings** page, hidden under the `Advanced` section.

<img width="760" alt="image" src="https://user-images.githubusercontent.com/3295/202047026-f4b10172-fb6f-4d23-a1a1-2b6d7a1a80c7.png">









#### Testing instructions





- Checkout this branch, run `npm run build:client`
- Open **Payments > Settings** and expand the **Advanced** section, see that there's a new toggle, labeled with **Enable UPE public key encryption**, and it should be **off** as default.
- Test if the toggle persists its value to `wp_options` table on the client database, with the option name `_wcpay_feature_client_secret_encryption`.
- Toggle the feature on.

 **Client Secret Encryption** | **Enabled** | **Enabled** | **Disabled** | **Disabled** 
------------------------------|-------------|-------------|--------------|--------------
 **UPE**                      | Enabled     | Disabled    | Enabled      | Disabled     
 **Regular sale**.            |             |             |              |              
 **Subscription renewal**     |             |             |              |
 **Free item sale**           |             |             |              |              
 **Save cards to account**    |             |             |              |              

And see all those work. Also while doing those; 
- check if the `wc-ajax=create_payment_intent` and `wc-ajax=create_setup_intent` endpoints return a encrypted version of the `client_secret` (not starting with the payment intent ID) when the encryption is enabled, 
- and if they return the plain `client_secret` when the encryption is disabled.

> **Note**: Free item sale is done like this: 
>
> Create a free trial subscription product first;
> - Product data > Simple Subscription
> - Subscription price > 10$ every month (any price)
> - Free trial > 30 days (any number of days)
> 
> Add the product to the cart, sell it.

-------------------


**Post merge**



- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-5.1.0#client-secret-encryption-toggle
